동기화 메모장 [Step2] 앨리, 제인, 조이

na-young-kwon opened the pull request:
안녕하세요 수박 🍉 
@soo941226
앨리, 제인, 조이 입니다.

말씀드린대로 Step1,2 모두 구현하여 pr 보냅니다
부족한 부분이 많지만 피드백 주시면 같이 고민해보겠습니다!
사소한 부분이라도 짚어주시면 감사하겠습니다.

리뷰 잘 부탁드립니다!😄

# STEP 1 : 리스트 및 메모영역 화면 UI구현

리스트 화면과 메모영역 화면을 SplitViewController를 활용하여 구현합니다.

## 1-1 고민했던 것

### 양방향 델리게이트로 메모 목록과 상세페이지간 데이터 전달

메모 목록을 테이블뷰 형식으로 가지고있는 `MemoListViewController` 와, 메모의 내용을 표시하는 `memoDetailViewController` 간에 데이터 전달을 위하여 양방향으로 delegation 관계를 구현하였습니다.

`MemoListViewController`는 테이블뷰 셀이 선택되면 UITableViewDelegate 메서드 didSelectRowAt에서 `MemoDetailViewControllerDelegate` 프로토콜을 채택한 `memoDetailViewController` 에게 Memo 인스턴스를 전달하여 텍스트뷰에 표시할 데이터를 전달합니다.

`memoDetailViewController`는 사용자가 메모의 내용을 수정하면 `MemoListViewControllerDelegate`을 채택한 `MemoListViewController` 에게 변경된 텍스트뷰의 내용을 전달하여 수정사항을 메모 목록에 반영합니다.

### NSAttributedString과 defaultContentConfiguration을 이용한 테이블 뷰 셀 구성

<img src="https://i.imgur.com/JXo1jzg.png" width="50%" height="50%">

- subtitle에서 날짜와 메모 본문에 다른 attribute를 적용하기 위해 NSAttributedString을 사용했습니다
- 날짜는 footnote, 본문은 caption1 + secondaryLabel 색상으로 구성했습니다.

```swift
let attributedString = NSMutableAttributedString()

attributedString.append(NSAttributedString(
    string: dateString + " ",
    attributes: [.font: UIFont.preferredFont(forTextStyle: .footnote)]
))

attributedString.append(NSAttributedString(
    string: truncatedBody,
    attributes: [
        .font: UIFont.preferredFont(forTextStyle: .caption1),
        .foregroundColor: UIColor.secondaryLabel
    ]
))
```

### UISwipeActionsConfiguration 사용

TableView의 Cell을 swipe할 때 공유 및 삭제 기능을 위한 액션버튼이 띄워지도록 구현했습니다.

<img src="https://i.imgur.com/n5FNemO.png" width="50%" height="50%">

### 어플의 실행에 따른 selectRow(at:) 호출

- 앱이 처음 구동될 때 첫번째 셀이 선택되도록 했습니다.
- 메모를 추가했을 때 추가한 새로운 메모를 select 합니다.
- 메모를 삭제했을 때는 삭제한메모의 다음 메모를 자동으로 select 합니다.

어떤 메모를 선택해서 작성하고 있는지 알리기 위해 작성중인 셀이 계속 select 되도록 구현했습니다.

### 키보드의 text 가림현상 개선

`NotificationCenter`를 활용하여 키보드가 화면에 표시될 때, textView의 text를 가리지 않도록 contentInset을 키보드의 높이와 같게 조정하고, `textView.isEditable`을 사용하여 메모가 없을 때 textView를 수정할 수 없도록 구현했습니다.

<img src="https://user-images.githubusercontent.com/74536728/154416221-22c394ea-9025-4e51-b4e7-f19d7dae4cf5.gif" width="50%" height="50%">

### 안전하게 배열 조회

배열에서 존재하지않는 인덱스를 조회했을 때 Crash가 나지 않도록 subscript를 활용하여 안전하게 조회할 수 있도록 했습니다.

```swift
extension Collection {
    subscript (safe index: Index) -> Element? {
        return indices.contains(index) ? self[index] : nil
    }
}
```

## 1-2 Trouble Shooting

### 1. GestureRecognizer의 cancelsTouchesInView 기본값 true

- GestureRecognizer를 ViewController에 추가하자 UITableView의 셀이 터치되지 않는 현상이 나타났습니다.
- `원인` 그 이유는 GestureRecognizer의 프로퍼티 cancelsTouchesInView의 기본값이 `true`이기 때문입니다. 제스처만 인식한 후 나머지 터치정보들을 뷰로 전달하지 않고 취소하기 때문에 UITableView의 UITableViewDelegate 메서드가 작동하지 않았습니다.
- `해결` 따라서 cancelsTouchesInView값을 `false`로 할당해줌으로써 제스처를 인식한 후에도 Gesture Recognizer와는 무관하게 터치 정보들을 뷰에 전달할 수 있게 되었습니다.

### 2. 셀 선택이 유지되지 않는 문제

|수정 전|수정 후|
|:---:|:---:|
|<img src="https://i.imgur.com/6nMVmAh.gif" width="50%" height="50%">|<img src="https://i.imgur.com/bhur7K6.gif" width="50%" height="50%">|

- `원인` tableView의 `allowsSelectionDuringEditing` 프로퍼티의 디폴트가 false였기 때문에 셀선택이 되지않았습니다.
- `해결`  `allowsSelectionDuringEditing` 를 true로 바꿔주었습니다. 셀을 지운 후에도 셀선택이 남아있도록 하기 위해 `didEndEditingRowAt`에서도 indexPath에 해당하는 row를 select하는 로직을 추가했습니다.
    
    ```swift
    // 셀을 지우는 동안 editing을 할 수 있도록 true로 변경
    tableView.allowsSelectionDuringEditing = true
    tableView.selectRow(at: indexPath, animated: false, scrollPosition: .top)
    
    // 셀 수정이 끝난 후에도 셀을 select하는 로직 추가
    func tableView(_ tableView: UITableView, didEndEditingRowAt indexPath: IndexPath?) {
      guard let indexPath = indexPath else {
          return
      }
      tableView.selectRow(at: indexPath, animated: false, scrollPosition: .top)
    }
    ```
    

### 3. 마지막 셀을 지웠을 때 index 오류가 나는 문제

```swift
Thread 1:
"Attempted to scroll the table view to an out-of-bounds row (0) when there are only 0 rows in section 0.
Table view: <UITableView: 0x13f031400;
frame = (0 0; 420 834);
clipsToBounds = YES;
autoresize = W+H; gestureRecognizers = <NSArray: 0x600000031680>;
layer = <CALayer: 0x600000ec7b80>; contentOffset: {0, -74};
contentSize: {420, 72.5}; adjustedContentInset: {74, 0, 20, 0};
dataSource: <CloudNotes.MemoListViewController: 0x14880fad0>>"
```

- `원인`  셀을 지운 후 다음 셀을 select해주는데, 맨 마지막 셀을 지우면 select할 row가 남아있지 않아서 index 오류가 발생했습니다.
- `해결`  조건문으로 지우려는 셀의 row가 남아있는 메모의 개수보다 작을때만 `tableView.selectRow(at: indexPath)` 를 하도록 분기해주었습니다. 추가로 맨 마지막 셀을 지우는 경우 detailViewController가 비어있는 텍스트뷰를 보여주도록 구현했습니다.

```swift
private func deleteMemo(at indexPath: IndexPath) {
    let deletedMemo = MemoDataManager.shared.memos[indexPath.row]
    MemoDataManager.shared.memos.remove(at: indexPath.row)
    tableView.deleteRows(at: [indexPath], with: .none)
    MemoDataManager.shared.deleteMemo(id: deletedMemo.id)
    
    if indexPath.row < MemoDataManager.shared.memos.count {
        let memo = MemoDataManager.shared.memos[indexPath.row]
        delegate?.memoDetailViewController(showTextViewWith: memo)
        tableView.allowsSelectionDuringEditing = true
        tableView.selectRow(at: indexPath, animated: false, scrollPosition: .top)
    } else {
		// 맨 마지막 셀을 지우는 경우
        delegate?.showEmptyTextView()
    }
}

func showEmptyTextView() {
    textView.isEditable = false
    textView.text = ""
}
```


# STEP 2 : 코어데이터 DB 구현

메모를 위한 코어데이터 모델을 생성합니다.

## 2-1 고민했던 것

### 코어데이터를 관리하는 매니저 타입 구현

- 데이터를 관리하는 타입인 `MemoDataManager` 를 구현했습니다.
    - fetch해온 데이터를 저장해 놓는 memos 배열을 가지고 있습니다.
    - CoreData를 생성, 삭제, 조회, 업데이트를 담당합니다.

### 사용자 친화적인 UI 구현

메모가 추가되거나 수정될 때 최신순으로 정렬하여 TableView에 보여줄 수 있도록 날짜를 기준으로 정렬했습니다.

메모리스트 뷰에서는 Swipe해서 share 버튼을 눌렀을 때, UIActivityView를 화면 중앙에 보여주도록 해주었고, 상세페이지의 barButtonItem에서 share 버튼을 눌렀을 때는, 해당 버튼에서부터 UIActivityView가 보여지도록 했습니다.

|Swipe해서 share 버튼을 눌렀을 때|barButtonItem에서 share 버튼을 눌렀을 때|
|:---:|:---:|
|<img src="https://i.imgur.com/Yn4FJqW.png" width="50%" height="50%">|<img src="https://i.imgur.com/28ZUpZF.png" width="50%" height="50%">|

### 아이패드에서 popoverPresentationController의 사용

- 오류메세지

아이폰에서는 잘 띄워지던 UIAlertController나 UIActivityViewController가 아이패드 환경에서는 작동하지 않았습니다. 

```
Thread 1: "Your application has presented a UIAlertController (<UIAlertController: 0x10d813a00>) of style UIAlertControllerStyleActionSheet from CloudNotes.SplitViewController (<CloudNotes.SplitViewController: 0x11f7068f0>).
The modalPresentationStyle of a UIAlertController with this style is UIModalPresentationPopover.
You must provide location information for this popover through the alert controller's popoverPresentationController.
You must provide either a sourceView and sourceRect or a barButtonItem.
If this information is not known when you present the alert controller, you may provide it in the UIPopoverPresentationControllerDelegate method -prepareForPopoverPresentation."

```

popoverPresentationController를 사용하여 얼럿이 띄워질 위치를 sender나 뷰의 특정한 위치로 명시를 하여 해결했습니다.

- sender를 준 경우

```swift
private func showAlert(_ sender: UIBarButtonItem) {
    let alert = UIAlertController(title: nil, message: nil, preferredStyle: .actionSheet)
    // 코드 생략
    if let popoverController = alert.popoverPresentationController {
        popoverController.barButtonItem = sender
    }
    present(alert, animated: true, completion: nil)
}
```

- 좌표를 설정해준 경우

```swift
private func showActivityView(indexPath: IndexPath) {
    //코드 생략
    let activityViewController = UIActivityViewController(activityItems: [memoToShare], applicationActivities: nil)
    
    if let popOver = activityViewController.popoverPresentationController {
        popOver.sourceView = splitViewController.view
        popOver.sourceRect = CGRect(x: splitViewController.view.bounds.midX,
                                    y: splitViewController.view.bounds.midY,
                                    width: 0,
                                    height: 0)
        popOver.permittedArrowDirections = []
    }
    present(activityViewController, animated: true)
}
```

<img src="https://i.imgur.com/1Dv8067.png" width="50%" height="50%">

[관련 공식문서 링크: Displaying Transient Content in a Popover](https://developer.apple.com/documentation/uikit/windows_and_screens/displaying_transient_content_in_a_popover)

## 2-2 Trouble Shooting

### UITableView의 Cell을 deleteRows(at:) 메서드로 삭제했을 때 발생한 에러

```
Thread 1:
"Invalid update: invalid number of rows in section 0.
The number of rows contained in an existing section after the update (26) must be equal to the number of rows contained in that section before the update (26), plus or minus the number of rows inserted or deleted from that section (0 inserted, 1 deleted) and plus or minus the number of rows moved into or out of that section (0 moved in, 0 moved out).
```

- `상황` tableView 섹션의 row 개수와 실제 보여줄 섹션의 개수가 맞지 않아서 발생하는 오류입니다.
- `이유` tableView의 셀을 삭제하면서 tableView에 보여줄 데이터도 같이 삭제 해주어야 하는데, 그 과정이 누락이되어 발생한 것으로 보입니다.
- `해결` 셀을 추가 및 삭제할 때 tableView에 보여줄 섹션의 개수도 동일하도록 `MemoDataManager`의 메모와 코어데이터에 저장된 메모데이터도 삭제 해주었습니다.

### 텍스트뷰가 변할때마다 메모가 저장되는 문제

|작성 중일때|앱을 다시 실행해서 코어데이터에서 fetch해왔을 때|
|:---:|:---:|
|<img src="https://i.imgur.com/yzjlrDt.png" width="50%" height="50%">|<img src="https://i.imgur.com/3ULs2Tt.png" width="50%" height="50%">|



- `상황` 텍스트뷰에 여러 글자를 적고, 앱을 다시 구동하면 오른쪽 사진처럼 한글자씩 메모가 여러개 생기는 오류입니다.
- `이유` 텍스트뷰가 수정이 될 때마다 코어데이터에 저장이 되기 때문에 발생한 문제였습니다.
    - DetailViewController에서 `currentMemo` 라는 연산프로퍼티로 현재 메모에 접근했는데,
    - 그 안에서 만들어주는 Memo 타입을 이니셜라이즈 할때 코어데이터의 context가 쓰이면서 글자를 수정한 수만큼 메모가 저장이 되고,
    - fetch를 해올 때 한 글자씩 저장 된 모든 메모들을 불러오면서 하나의 메모를 수정했지만, 수정한 글자 수대로 새로운 메모들이 생기는 현상이 발생하였습니다.

<수정 전>

```swift
extension MemoDetailViewController: UITextViewDelegate {
		private var currentMemo: Memo {
        let memoComponents = textView.text.split(
            separator: "\n",
            maxSplits: 1
        ).map(String.init)
        
        **let memo = Memo(context: MemoDataManager.shared.viewContext)**
        memo.title = memoComponents[safe: 0] ?? ""
        memo.body = memoComponents[safe: 1] ?? ""
        memo.lastModified = Date()
        
        return memo
    }

    func textViewDidChange(_ textView: UITextView) {
        delegate?.memoListViewController(updateTableViewCellWith: currentMemo)
    }
}

```

프린트문으로 출력해본 결과 `context`의 수는 글자가 입력할 때 마다 늘어나지만, `메모`의 수는 변함이 없는것을 확인할 수 있었습니다.

<img src="https://i.imgur.com/If3slj0.png" width="20%" height="20%">

```swift
extension MemoListViewController: MemoListViewControllerDelegate {
		func memoListViewController(updateTableViewCellWith memo: Memo) {
		    guard let indexPath = tableView.indexPathForSelectedRow else {
		        return
		    }
		    let request = NSFetchRequest(entityName: "Memo")
				//context의 수가 글자 입력할때마다 늘어남 
		    print(try? MemoDataManager.shared.viewContext.count(for: request))
				
		    MemoDataManager.shared.memos[indexPath.row] = memo
				//메모의 수는 그대로 
		    print(MemoDataManager.shared.memos.count)
		    tableView.reloadRows(at: [indexPath], with: .none)
		    tableView.selectRow(at: indexPath, animated: true, scrollPosition: .none)
		}
}
```

- `해결`  코어데이터의 `Memo(context:)` 객체를 이용하지 않고 그냥 메모 수정 화면에서 메모 목록으로 title, body를 자체를 전달하여 수정되고 있는 셀의 indexPath로 MemoDataManager에서 Memo 를 가져와서 MemoDataManager.shared.updateMemo 로 해당 Memo를 수정한 후 context를 save하는 방식으로 해결했습니다.

```swift
extension MemoDetailViewController: UITextViewDelegate {
		func textViewDidChange(_ textView: UITextView) {
        let memoComponents = textView.text.split(separator: "\n",
                                                 maxSplits: 1)
                                                .map(String.init)
        
        let title = memoComponents[safe: 0] ?? ""
        let body = memoComponents[safe: 1] ?? ""
        let lastModified = Date()
        
        delegate?.memoListViewController(updateTableViewCellWith: title, body: body, lastModified: lastModified)
    }
}

extension MemoListViewController: MemoListViewControllerDelegate {
		func memoListViewController(updateTableViewCellWith title: String, body: String, lastModified: Date) {
        guard let indexPath = tableView.indexPathForSelectedRow,
              let id = MemoDataManager.shared.memos[indexPath.row].id else {
            return
        }
        MemoDataManager.shared.updateMemo(id: id, title: title, body: body, lastModified: lastModified)
        tableView.reloadRows(at: [indexPath], with: .none)
        tableView.selectRow(at: indexPath, animated: true, scrollPosition: .none)
    }
}

class MemoDataManager {
		func updateMemo(id: UUID,
                    title: String,
                    body: String,
                    lastModified: Date)
		    {
        let predicate = NSPredicate(format: "id == %@", id.uuidString)
        guard let memo = fetchMemos(predicate: predicate).first else {
            return
        }
        memo.title = title
        memo.body = body
        memo.lastModified = lastModified
        saveViewContext()
    }
}
```

### 마지막 Cell을 삭제했을 때, textView에 text가 남아있는 문제

MemoListViewController의 `deleteMemo()` 메소드 내부에서 조건문을 통해 마지막 Cell인지 확인 후 detailViewController에 있는 `showEmptyTextView()` 메소드를 호출해 textView의 text를 초기화 해주었습니다.

```swift
private func deleteMemo(at indexPath: IndexPath) {
    let deletedMemo = MemoDataManager.shared.memos[indexPath.row]
    MemoDataManager.shared.memos.remove(at: indexPath.row)
    tableView.deleteRows(at: [indexPath], with: .none)
    MemoDataManager.shared.deleteMemo(id: deletedMemo.id)
        
    if indexPath.row < MemoDataManager.shared.memos.count {
        let memo = MemoDataManager.shared.memos[indexPath.row]
        delegate?.memoDetailViewController(showTextViewWith: memo)
        tableView.allowsSelectionDuringEditing = true
        tableView.selectRow(at: indexPath, animated: false, scrollPosition: .top)
    } else {
        delegate?.showEmptyTextView()
    }
}
```

na-young-kwon committed:
style: swiftLint 적용

siwonkim0 committed:
feat: Storyboard 삭제 후 SceneDelegate에서 SplitViewController를 root로 설정 #1

- UISplitViewController 타입의 SplitViewController 생성하여 초기 세팅

Allie-e committed:
feat: Memo, MemoDataManager 구현 #3

- Memo extension으로 attributedString을 사용하여 body와 lastModified를 하나의
문자열로 병합
- String extension으로 문자열의 길이를 제한하는 truncated() 메소드 구현
- MemoDataManager 타입을 구현하여 JSON data를 파싱하여 가지도록 함

na-young-kwon committed:
feat: MemoListViewController의 테이블뷰 구현 #1

- UITableViewDatasource 구현
- NavigationBar구현
- 메모 추가 기능 구현
- MemoDataManager의 shared프로퍼티 static으로 변경
- Memo의 subtitle 글자제한을 30 -> 40으로 변경

siwonkim0 committed:
feat: MemoDetailViewController의 textView, List->Detail 데이터 전달 구현 #2

- MemoDetailViewControllerDelegate 생성
- List뷰컨의 delegate를 detail 뷰컨으로 지정하여 detail뷰컨의 textView 업데이트
- List뷰컨에서 새로운 메모 추가시 제일 첫번째 셀이 선택되도록 구현

Allie-e committed:
feat: MemoListViewControllerDelegate 구현 #2

- detailVC의 delegate를 listVC으로 지정하여 textView에 수정이 있을 때 cell에
반영하도록 구현
- add 버튼 클릭했을 때, 첫번째 cell이 선택되도록 수정
- collection extension 추가하여 안전하게 배열 조회

na-young-kwon committed:
feat: 메모 삭제 기능 구현 #2

- tableView Delegate의 메서드인 trailingSwipeActionsConfigurationForRowAt을 사용하여 셀을 삭제하도록 구현
- didEndEditingRowAt 메서드를 통해 셀삭제 후에도 셀선택을 유지하도록 구현

siwonkim0 committed:
feat: keyboard 관련 기능 구현 #2

- NotificationCenter를 활용하여 textView에 맞게 키보드를 올리는 기능 구현
- UITapGestureRecognizer를 활용하여 빈 뷰를 선택하면 키보드를 내리는 기능 구현

Allie-e committed:
feat: CoreData 적용 #4

- JSON Data Model 및 관련 코드 삭제
- MemoDataManager의 persistentContainer 생성 및 saveViewContexte() 메소드 구현

na-young-kwon committed:
fix: 맨 마지막 셀 삭제시 index오류가 나는 버그 수정 #6

- indexPath와 메모의 개수를 비교하여 indexPath가 작거나 같을때만
 셀을 삭제하도록 수정
- deleteMemo 함수 분리
- title, body가 모두 비어있을 때 textView.text를 emptyString으로 수정

siwonkim0 committed:
feat: Detail 뷰컨의 더보기 버튼에 ActionSheet 형태의 Alert 구현 #5

- popoverPresentationController를 사용하여 아이패드 환경에서도 Alert이 띄워지도록 구현

Allie-e committed:
feat: viewMoreButtonTapped() 세부사항 구현 #5

- share 버튼 눌렀을 때, activityView 보여지도록 추가
- delete 버튼 눌렀을 때, alert 보여지도록 추가

na-young-kwon committed:
feat: 메모셀을 스와이프했을 때 공유버튼 구현 #5

- 공유버튼을 누르면 activityView가 스플릿뷰의 중심에 나타나도록 구현

siwonkim0 committed:
fix: 메모 수정 중에 코어데이터에 여러번 업데이트되는 현상 수정 #6

- 기존에는 코어데이터의 entity인 Memo 타입을 텍스트뷰가 수정될때마다 업데이트하였으나,
viewContext가 여러번 저장되어 title, body, lastModified만 업데이트하는 방식으로 수정

Allie-e committed:
feat: MemoDataManager의 deleteMemo() 메소드 추가 #4

- MemoDetailViewControllerDelegate에 showEmptyTextView() 메소드 추가하여 마지막
셀 삭제될 때 빈 textView를 보여주도록 구현

na-young-kwon committed:
feat: 더보기 버튼을 눌렀을 때 셀을 삭제하는 기능 구현 #5

- MemoListViewControllerDelegate에 deleteTableViewCell메서드 추가
- 삭제버튼을 눌렀을 때 delegate가 테이블뷰 셀을 삭제하도록 구

 Please enter the commit message for your changes. Lines starting

siwonkim0 committed:
refactor: MemoDataManager 접근제어 추가, 로직 수정 #7

Allie-e committed:
refactor: MemoListViewController 전제척인 refactoring #7

na-young-kwon committed:
refactor: MemoListViewController 전체적인 refactoring #7

- 전체적으로 final 키워드 붙임
- 외부로 노출할 필요가 없는 메서드 private 메서드로 변경
- 마지막 메모가 삭제된 후 textView의 수정이 불가능한 버그 수정

siwonkim0 committed:
refactor: MemoDetailViewController 메서드 분리 및 재배치 #7

Allie-e committed:
refactor: 호출순서에 따른 코드 위치 수정 및 refactoring #7

na-young-kwon committed:
refactor: MemoDataManager의 deleteMemo() 메서드 수정

- 인덴트 수정
- memo -> memos로 네이밍 변경
- 타입명시 삭제

siwonkim0 committed:
docs: STEP 1, 2 README 작성

Allie-e committed:
docs: Update README.md

soo941226 commented on CloudNotes/CloudNotes.xcodeproj/project.pbxproj:
```diff
@@ -256,6 +261,27 @@
 		};
 /* End PBXResourcesBuildPhase section */
 
+/* Begin PBXShellScriptBuildPhase section */
+		3B69FE7A27BB3B37002B216E /* Swift Lint */ = {
+			isa = PBXShellScriptBuildPhase;
+			buildActionMask = 2147483647;
+			files = (
+			);
+			inputFileListPaths = (
+			);
+			inputPaths = (
+			);
+			name = "Swift Lint";
+			outputFileListPaths = (
+			);
+			outputPaths = (
+			);
+			runOnlyForDeploymentPostprocessing = 0;
+			shellPath = /bin/sh;
+			shellScript = "if test -d \"/opt/homebrew/bin/\"; then\n  PATH=\"/opt/homebrew/bin/:${PATH}\"\nfi\n\nexport PATH\n\nif which swiftlint >/dev/null; then\n  swiftlint\nelse\n  echo \"warning: SwiftLint not installed, download from https://github.com/realm/SwiftLint\"\nfi\n";
+		};
+/* End PBXShellScriptBuildPhase section */
```
SwiftLint를 적용하셨군요! 👍 정말 좋습니다. 그런데 어떤 의존성 관리도구를 선택하신 것인지 저는 파악하지 못했는데요.
따로 선택하신 의존성 관리도구가 있으신 건가요?
혹시 의존성 관리도구를 선택하지 않으셨다면 왜 이러한 것을 거치지 않고 SwiftLint를 적용하셨는지 궁금하네요!

na-young-kwon replied:
따로 의존성 관리 도구를 사용하지 않고 HomeBrew를 통해 SwiftLint를 설치했습니다.
이렇게 하면 프로젝트마다 따로 SwiftLint를 적용하지 않고 전역적으로 사용할 수 있다는 장점이 있는 것 같아요.

soo941226 replied:
오 한 사람이 혼자 많은 프로젝트를 담당한다면 말씀하신 게 분명한 장점을 가질 거 같아요! 납득했습니다! 정말 좋네요👍
 그런데 지금과 같은 경우에는 세분이 하나의 프로젝트를 담당하는 건데, 세분의 SwiftLint가 버전이 다르다거나 하는 등의 이유로, 서로 다른 경고와 에러를 보게 된다면 일관된 컨벤션을 유지하기 어려워질 수 있겠네요. Lint 버전이야 맞추면 되긴 하지만, 또 그렇다고 한다면 사람이 그런 걸 일일히 확인을 하고 맞추는 것보단 도구에게 일을 시키고 도구가 알아서 일을 해줄 수 있는데 굳이 사람이 해야할까? 라는 생각도 들긴 하네요.
 의존성 관리도구를 쓰라는 이야기는 아니지만, 협업을 고려했을 때에는 단점으로 작용하는 부분이 아닐까 하는 생각이 들긴 하네요😅

na-young-kwon replied:
오,, 그런부분까진 생각 못해봤어요
버전의 차이 때문에 발생하는 문제 , 버전관리를 대신 해준다는 장점이 있겠네요
확실히 의존성 관리도구를 사용하는것이 좋겠어요. 좋은 피드백 감사합니다👍🏻👍🏻

soo941226 commented on CloudNotes/CloudNotes/SplitViewController.swift:
```diff
@@ -0,0 +1,23 @@
+        preferredDisplayMode = .oneBesideSecondary
+    }
+    
+}
```
접근제어를 붙여주신 게 참 좋네요👍

그런데 setUpDisplay라는 이름은 적절하다는 생각이 들면서도, setUpChildView는 이름과 내용이 적절히 매칭이 되지 않는다는 생각도 드네요!

그리고 스위프트의 퍼포먼스와 관련된 내용을 공부하신 것으로 알고 있는데요! static dispatch와 dynamic dispatch가 무엇인지 기억하고 계실까요? 이러한 점을 고려하여서 현재의 viewDidLoad가 런타임에 어떻게 실행이 될지 고민을 해보시면 좋을 거 같습니다!

Allie-e replied:
childView를 등록한다는 의미에서 `setUpChildView()`로 네이밍해주었는데 다른 관점에서 보면 childView에 대한 어떤 설정을 해준다는 의미로 받아들일 수 있을 것 같아요! `setViewController()` 메소드에대한 공식문서를 보니 아래와 같이 assign이라는 표현을 사용하고 있어, `assignChildViewController()` 로 네이밍 수정해주었습니다.
> For a column-style split view controller, you use this method to assign child view controllers to a specific column of the split view interface.

스위프트 퍼포먼스 관련해서 말씀해주신 내용은 class는 참조타입에 해당되어 항상 상속의 가능성에 노출되어있기 때문에 dynamic dispatch를 사용하는데 상속의 가능성이 없다는 것을 확실히 해준다면, 컴파일러가 static dispatch로 바꿔줄 수 있습니다. 따라서 final 키워드를 사용해 서브클래싱을 막아주어야한다는 뜻으로 이해했는데, 의도하신게 맞을까요?

soo941226 replied:
`assignChildViewController` 이름 정말 좋네요! 그런데 메소드 명은 하나의 뷰컨트롤러를 할당하는 것 같은데, 내부에서는 여러개의 뷰컨트롤러를 할당하게 되지 않나요?🤔

제 생각에는 아마 final을 붙이지 않아도 현재의 코드에서 viewDidLoad는 override를 했기 때문에 static dispatch로 동작을 하지 않았을까 하는 생각이 드는데요. 아마 다른 라이프사이클 메소드들은 dynamic dispatch가 되었을 것 같은데 잘 모르겠네요! 저도 WWDC를 한번 다시 봐야 기억이 생생히 날 것 같습니다. final을 의도한 건 맞습니다~!

soo941226 commented on CloudNotes/CloudNotes/AppDelegate/SceneDelegate.swift:
```diff
@@ -12,8 +12,8 @@ class SceneDelegate: UIResponder, UIWindowSceneDelegate {
+        window?.rootViewController = splitViewController
         window?.makeKeyAndVisible()
     }
 
```
풀어서 써주시는 거 정말 좋아요~👍

soo941226 commented on CloudNotes/CloudNotes/Model/Memo.swift:
```diff
@@ -0,0 +1,35 @@
+        ))
+        return attributedString
+    }
+}
```
NSAttributedString을 사용하셨네요! 정말 좋은 시도인 거 같아요👍
그런데 날짜를 표현하는 부분과 바디의 일부분을 표현하는 부분이 같이 정의가 되어있네요. subtitle이라는 것 하나로 컨트롤러에서 사용하는 것은 좋다고 생각되지만, subtitle안에서 이렇게 많은 내용이 정의가 되는 건 피하는 게 좋을 거 같아요.
지금은 그나마 20줄 안팎이라 읽는데 문제가 없지만, 이게 점점 옵션이 추가되어 길어지다가 문제가 터지면, 어떤 부분을 고쳐야하는지 처음부터 읽는 게 노동이 될 수도 있을 거 같네요!

siwonkim0 replied:
말씀해주신 것처럼 subtitle에서 너무 많은 일을 하고 있다고 생각합니다. 따라서 날짜를 표현하는 부분을 따로 메서드로 분리해주었습니다. 바디의 일부분을 표현하는 부분은 한줄이라서 따로 메서드로 분리하진 않았습니다..!

soo941226 commented on CloudNotes/CloudNotes/Model/MemoDataManager.swift:
```diff
@@ -0,0 +1,25 @@
+            print(error)
+        }
+    }
+}
```
먼저 이 부분은 실행이 되는지 테스트가 된 후에 커밋이 된 게 맞을까요? 제 생각에는 init이 재귀가 되면서 앱이 멈추지 않았을까 하는 생각이 드는데요🤔

그리고 이건 이미 지나버린 일이지만, loadJSON이라고 하는 내용도 이 당시 CoreData를 사용하는 것을 염두해두고 만드셨더라면 좋지 않았을까 하는 아쉬움이 남네요. 저는 JSON을 가져오는 내용과 CoreDataModel을 가져오는 내용이 적절히 모두 관리가 되면 좋겠습니다.

na-young-kwon replied:
<img width="1411" alt="스크린샷 2022-02-21 오후 4 54 33" src="https://user-images.githubusercontent.com/74536728/154911720-8ee89528-34ff-47c7-a8f9-3ceb07e16ddf.png">

예상과는 다르게 정상적으로 실행이 되고 있는것을 확인했어요. 
앱이 멈춘다거나 하는 문제도 찾지 못했구요,,ㅜㅜ
혹시 위 같이 작성했을 때 수박이 우려하신 상황이 어떤것인지 좀 더 자세히 알 수 있을까요?

둘은 완전 별개라고 생각해서 JSON을 가져오는 부분과 CoreDataModel을 가져오는 부분을 관리 할 생각을 못했던 것 같아요. 
관련해서 방법이 잘 떠오르지 않는데, 이것도 수박의 의견을 한번 여쭤봐도 될까요?

soo941226 replied:
일단 조이가 올려주신 코드는 현재의 코드인 것 같네요! 제가 코멘트를 남긴 코드는 꽤 이전의 커밋입니다. 해당 커밋에서는 `shared`가 타입 프로퍼티가 아니였어요!
```swift
class MemoDataManager {
    let shared = MemoDataManager()
}
```
위와 같은 구조였는데 `MemoDataManager`가 다시 `MemoDataManager`를 가지고 있기 때문에 무한 init에 빠지는 것으로 읽혔습니다!

아래의 질문에 대해서는 그냥 함수를 두 개 두고 init을 할 때 어떤 상태를 주입받아서, 그 상태에 따라 json을 load할 것인지 CoreData를 load할 것인지를 결정하면 될 것 같다는 생각이 드네요!

soo941226 commented on CloudNotes/CloudNotes/Controller/MemoListViewController.swift:
```diff
@@ -1,8 +1,59 @@
+            tableView.trailingAnchor.constraint(equalTo: view.trailingAnchor),
+            tableView.bottomAnchor.constraint(equalTo: view.bottomAnchor)
+        ])
+    }
```
tableView의 dataSource와 셀을 등록하는 부분은 setup 바깥에 두신 이유가 있을까요?

Allie-e replied:
기존에는 dataSource와 셀을 등록하는 부분은 viewDidLoad()에 있고 레이아웃 관련된 코드는 `setupTableViewLayout()` 메소드 안에 있었습니다. 리팩토링을 하면서 dataSource와 셀을 등록하는 부분을 `setupTableView()` 메소드로 빼주었습니다. 특별한 이유보다는 레이아웃 관련된 코드와 분리를 해주고 싶었습니다 😊

<img width="1054" alt="image" src="https://user-images.githubusercontent.com/83864058/154911896-68543769-5c18-4b8a-896a-1a01a6c92da1.png">

soo941226 replied:
정말 좋네요~~👍
그런데 `setupTableView`와 `setupTableViewLayout`이 같이 있으니까 이 둘을 호출하는 곳에서는 `setupTableView`는 뭘 하는 거지라고 확인을 해야만 할 것 같아요. `setupTableViewLayout`처럼 무엇을 setup하는 건지 표현이 되면 정말 좋을 거 같아요~

Allie-e replied:
이부분에 대한 코멘트는 밑에 작성했으니 참고 부탁드립니다~!

soo941226 commented on CloudNotes/CloudNotes/Controller/MemoListViewController.swift:
```diff
@@ -1,8 +1,59 @@
+        
+        cell.contentConfiguration = configuration
+        cell.accessoryType = .disclosureIndicator
+        return cell
```
`cell.defaultContentConfiguration`을 활용하셨네요! 정말 좋네요👍 그런데 해당 메소드가 iOS의 몇버전부터 지원을 하는지 혹시 염두해두셨을까요? 그렇다고 한다면 프로젝트가 지원하는 iOS 버전이 왜 현재의 값으로 설정이 되었는지 궁금합니다!

Allie-e replied:
`cell.defaultContentConfiguration` 메소드는 iOS 14이후 버전부터 지원을 하는 것으로 알고 있습니다! 다만, 현재 저희 프로젝트 타깃 버전을 확인해 본 결과 아래와같이 14.3으로 확인 되었습니다만, 저희가 14.3을 특정해서 설정해준 것은 아니고 자동으로? 설정된 버전인 것 같아요,,ㅎㅎ,,,

<img width="335" alt="image" src="https://user-images.githubusercontent.com/83864058/154912656-a3886810-c05a-4e9d-a351-64d3a177e939.png">

soo941226 replied:
아하 그렇군요. 지금은 아무런 상관이 없긴 하지만, 저는 프로젝트의 요구사항에 맞추어서 기능이 구현되어야 한다고 생각을 하는데요. 이렇게 생각을 하면서도, 굳이 그럴 필요가 없는 것이라면 기능에 맞게 이러한 버전을 낮추는 게 조금 더 많은 유저들이 사용할 수 있기 때문에 좋지 않을까라는 생각이 드네요. 기능구현을 어떻게 할 지 고민하는 요소가 될 수 있을 것 같아서 이야기를 드려봤습니다~

Allie-e replied:
deployment target 14.0으로 조정해주었습니다 😊

https://github.com/na-young-kwon/ios-cloud-notes/commit/0db139f46d9fdcc49249d218879278247305a62e

soo941226 commented on CloudNotes/CloudNotes/Model/Memo.swift:
```diff
@@ -14,7 +14,7 @@ extension Memo {
         
         let dateString = dateFormatter.string(from: lastModified)
-        let truncatedBody = body.truncated(limit: 30)
+        let truncatedBody = body.truncated(limit: 40)
```
매직넘버는 피하는 게 좋을 거 같아요~

siwonkim0 replied:
저희가 사전에 룰을 정했을때 두번이상 반복되는 매직넘버의 경우만 따로 enum으로 빼주기로 해서 이 부분은 한번이라 넘어갔습니다... 한번만 사용되는 경우에도 따로 처리를 해주는게 좋을까요? 👀

soo941226 replied:
처음부터 끝까지 그냥 30이었다면 그냥 그러려니 했을 것 같은데, 이렇게 값이 바뀌는 커밋이 있다면 왜 30이었고 왜 40이 나왔는지 저는 전혀 알 수가 없기 때문에 이름이 붙어야 한다고 생각을 합니다.

Allie-e replied:
넵 네임스페이스 만들어 관리해주도록 열거형 활용하여 수정해주었습니다~!

https://github.com/na-young-kwon/ios-cloud-notes/commit/0db139f46d9fdcc49249d218879278247305a62e

soo941226 replied:
너무 좋아요~~~~!!

soo941226 commented on CloudNotes/CloudNotes/Controller/MemoDetailViewController.swift:
```diff
@@ -1,8 +1,52 @@
+        textView.adjustsFontForContentSizeCategory = true
+        textView.translatesAutoresizingMaskIntoConstraints = false
+        return textView
+    }()
```
외부에서 접근해야하는 요소인가요?👀

Allie-e replied:
PR 제출할 때를 기준으로 private 키워드 붙여주었는데, 수박이 확인하신 부분이 이전 커밋인 것 같아요! 😊
앞으로는 코드를 작성할 때부터 접근제어자 신경써서 붙여주도록 하겠습니다~!

<img width="778" alt="image" src="https://user-images.githubusercontent.com/83864058/154913281-f989d627-6472-4636-b47a-c4e58fa0ee31.png">

soo941226 commented on CloudNotes/CloudNotes/Controller/MemoListViewController.swift:
```diff
@@ -3,10 +3,12 @@ import UIKit
 class MemoListViewController: UIViewController {
     private let cellIdentifier = "Cell"
     private let tableView = UITableView()
+    weak var delegate: MemoDetailViewControllerDelegate?
```
delegate를 활용하셨네요👍 정말 좋습니다.
다만 프로퍼티의 이름과 타입의 이름은 조금 더 고민을 해보시면 좋을 거 같아요.
예를 들어 tableView의 delegate는 UITableViewDeleagte인데,
memoListViewController의 delegate가 MemoDetailViewControllerDelegate인 게 저로써는 적절하지 않다는 생각이 드네요.

na-young-kwon replied:
넵
사장의 대리인은 직원이어야 하는데, 사장의 대리인이 직원의 대리인인 느낌,, 이라고 말씀해 주셔
MemoListViewController의 delegate가 MemoDetailViewController인게 부자연스럽다고 생각했습니다.

지금은 해당 프로토콜들을 없애는 쪽으로 개선완료했습니다.
delegate 패턴에 대해서 다시한번 알아보는 시간이었어요 감사합니다 :)

na-young-kwon replied:
넵
사장의 대리인은 직원이어야 하는데, 사장의 대리인이 직원의 대리인인 느낌,, 이라고 말씀해 주셔
MemoListViewController의 delegate가 MemoDetailViewController인게 부자연스럽다고 생각했습니다.

지금은 해당 프로토콜들을 없애는 쪽으로 개선완료했습니다.
delegate 패턴에 대해서 다시한번 알아보는 시간이었어요 감사합니다 :)

soo941226 commented on CloudNotes/CloudNotes/Controller/MemoDetailViewController.swift:
```diff
@@ -1,10 +1,12 @@
 }
 
 class MemoDetailViewController: UIViewController {
+    weak var delegate: MemoListViewControllerDelegate?
```
이 부분도 MemoDetailViewController의 delegate가 MemoListViewControllerDelegate인 것에 대해서 고민을 해보셨으면 좋겠습니다!

na-young-kwon replied:
이 부분도 위의 코멘트와 같이 좀 더 공부해보겠습니다~!

soo941226 commented on CloudNotes/CloudNotes/Extension/Collection+Extension.swift:
```diff
@@ -0,0 +1,5 @@
+extension Collection {
+    subscript (safe index: Index) -> Element? {
+        return indices.contains(index) ? self[index] : nil
+    }
+}
```
안전한 서브스크립트 너무 좋습니다👍💯

soo941226 commented on CloudNotes/CloudNotes/Controller/MemoListViewController.swift:
```diff
@@ -64,20 +64,42 @@ extension MemoListViewController: UITableViewDataSource {
+        }
+        deleteAction.image = UIImage(systemName: "trash")
+        return UISwipeActionsConfiguration(actions: [deleteAction])
+    }
```
deleteAction에게 넘겨주는 클로저는 이름을 붙여서 따로 관리를 하는 게 좋지 않을까요?

Allie-e replied:
해당 부분 따로 메서드로 작성하여 분리해주었습니다! 😊

<img width="1209" alt="image" src="https://user-images.githubusercontent.com/83864058/154917060-cb1710d9-c46d-470c-9323-de83ace51b51.png">

soo941226 commented on CloudNotes/CloudNotes/Controller/MemoDetailViewController.swift:
```diff
@@ -35,13 +35,32 @@ class MemoDetailViewController: UIViewController {
+
+        let keyboardRect = keyboardFrame.cgRectValue
+        textView.contentInset.bottom = keyboardRect.height
+        textView.scrollIndicatorInsets = textView.contentInset
```
느낌표가 있네요!

Allie-e replied:
수정해주었습니다!!ㅎㅎ

soo941226 commented on CloudNotes/CloudNotes/Controller/UIViewController+Extension.swift:
```diff
@@ -0,0 +1,19 @@
+    @objc func dismissKeyboard() {
+        view.endEditing(true)
+    }
+}
```
extension을 활용하신 것은 좋아요!
그런데 모든 UIViewController가 해당 기능을 갖는다는 건, 불필요한 정보가 생기는 UIViewController가 있다는 말 같아요. 이러한 부분을 신경써보시면 좋을 거 같습니다
또 hideKeyboard라는 이름과 안의 내용이 적절히 매칭이 되지 않는 것 같아요. 내용은 키보드를 감춘다기보다는 그러한 기능을 하는 tapGesture를 등록하는 것으로 읽히네요!

na-young-kwon replied:
확장성을 고려해서 UIViewController extension을 활용했는데, 
불필요한 정보가 생길수도 있다는 생각은 못했네요

말씀해주신대로 tap gesture를 등록하고 있어서 
hideKeyboard -> registerGestureRecognizer로 이름 변경해주었습니다.

soo941226 replied:
아.... `registerGestureRecognizer`가 너무 좋은데 약간 찝찝해서 이야기를 안 드릴 수가 없을 것 같습니다...! 무엇을 위한 어떤 GestureRecognizer가 등록이 되는 건가요?

na-young-kwon replied:
키보드를 숨기기 위한 tap gesture recognizer를 등록하는 작업입니다.
따라서
registerGestureRecognizerForHidingKeyboard
으로 변경했습니다

Allie-e replied:
음 `registerGestureRecognizerForHidingKeyboard` 로 수정해봤는데 괜찮을까요?ㅎㅎ

https://github.com/na-young-kwon/ios-cloud-notes/commit/0db139f46d9fdcc49249d218879278247305a62e

soo941226 replied:
너무 좋습니다!

soo941226 commented on CloudNotes/CloudNotes/Controller/MemoListViewController.swift:
```diff
@@ -17,7 +17,10 @@ class MemoListViewController: UIViewController {
+        
+        if !MemoDataManager.shared.isEmpty {
+            tableView.selectRow(at: IndexPath(row: 0, section: 0), animated: false, scrollPosition: .top)
+        }
```
저는 이렇게 MemoDataManager.shared.isEmpty를 체크를 해서 관리하게 되는 기능이 listViewController와 detailViewController에 분산되어 있다고 느껴집니다. 이 둘을 어디선가는 한번에 관리할 수 있지 않을까요?

siwonkim0 replied:
listViewController에서 MemoDataManager.shared.isEmpty 가 아닌 경우 (데이터가 있는 경우) 0번째 셀을 선택하는 것과, 0번째 데이터를 memoDetail에 보여주는 기능을 한번에 관리하도록 수정하였습니다~!

na-young-kwon replied:
이번에 대대적인 리팩토링을 진행하면서 MemoDataManager가 
조건문을 통해 listDelegate와 detailDelegate에게 적절한 일을 시키는 것으로 변경했습니다.
따라서 수박이 말씀하신 데이터를 한곳에서 관리하는 부분이 개선된것같습니다.

soo941226 commented on CloudNotes/CloudNotes/Model/MemoDataManager.swift:
```diff
@@ -1,25 +1,50 @@
 import UIKit
```
MemoDataManager가 UI요소인가요?

siwonkim0 replied:
CoreData가 Foundation을 참조하고 있는 관계라서 UIKit를 import 하는게 아예 필요가 없을 것 같아 삭제했습니다.

soo941226 commented on CloudNotes/CloudNotes/Model/MemoDataManager.swift:
```diff
@@ -1,25 +1,50 @@
+    let persistentContainer: NSPersistentContainer
+    var viewContext: NSManagedObjectContext {
+        return persistentContainer.viewContext
+    }
```
CoreData를 한문장으로 정리하면 어떤 걸까요? viewContext가 연산프로퍼티면 어떻게 될까요?

Allie-e replied:
공식문서를 보면, 디바이스에서 데이터를 저장 또는 캐시하거나 CloudKit을 사용하여 여러 디바이스에 데이터를 동기화하는 것이라고 나와있습니다.

viewContext가 연산프로퍼티일때 viewContext에 접근할 때마다 여러개의 viewContext가 생길수 있을 것 같아 확인을 해보니 아래와 같이 같은 주소를 참조하고 있어서 동작하는데 문제가 없다고 생각합니다! 질문의 의도를 저희가 잘 파악한 건지 모르겠어요 🤔!

<img width="400" alt="image" src="https://user-images.githubusercontent.com/83864058/154954083-2aa2795c-791b-4390-93ff-af19f66abde7.png">

soo941226 replied:
오 공식문서를 인용한 설명 너무 좋네요. 그런데 세분도 정말 그렇게 정리를 하고 계신 게 맞는지 조금 의문이 남긴 하네요🤔

viewContext는 NSManagedObjectContext 타입이고, 얘는 class이기 때문에 지금과 같은 동작을 하는 것인데요. 말씀하신 것처럼 문제는 없습니다. 제가 질문을 조금 생각없이 한 것 같네요😅 그런데 고민을 하신 내용을 보면 세분도 고민을 하시긴 한 것 같아요. viewContext에 접근할 때마다 연산프로퍼티가 연산을 해서 값을 반환해주기 때문에, 그만큼 참조에 대한 비용이 발생을 하게 될 것 같습니다. viewContext가 꼭 연산프로퍼티여야 할까요?🤔

Allie-e replied:
비용적인 부분에 대해서는 생각하지 못했습니다~ 그렇다면 연산프로퍼티보다는 저장프로퍼티로 가지고있는게 좋을 것 같아요! 수정해주었습니다😊

https://github.com/na-young-kwon/ios-cloud-notes/commit/0db139f46d9fdcc49249d218879278247305a62e

soo941226 commented on CloudNotes/CloudNotes/Model/MemoDataManager.swift:
```diff
@@ -1,25 +1,50 @@
+        persistentContainer.loadPersistentStores { description, error in
+            guard error == nil else {
+                fatalError(error!.localizedDescription)
+            }
```
음... 이 부분은 처음 탬플릿으로 제공하던 내용을 전혀 바꾸지 않았다고 생각되는데요. 제가 받아들이기엔 감자를 potato라고 쓴 느낌이에요. 탬플릿이 제공하는 주석에서는 어떤 이야기를 하고 있었나요?

na-young-kwon replied:
fatalError를 발생시키지 않고 에러를 적절하게 처리하라고 주석이 남아있는것을 확인했습니다.
(*Replace this implementation with code to handle the error appropriately. )

여기서 오류가 발생하는 원인은  아래와 같다고 합니다
* 상위 디렉터리가 없거나, 만들 수 없거나, 쓰기를 허용하지 않을 때
* 디바이스가 잠겨 있을 때 권한 또는 데이터 보호로 인해 영구 저장소에 액세스할 수 없을 때
* 장치에 공간이 부족할 때
* 저장소를 현재 모델 버전으로 마이그레이션할 수 없을 때

저희가 생각해본 방법으로는 
1. alert 띄우기 -> 아직 뷰가 만들어지기 전이라서 alert을 띄울 수 없음  (MemoDataManager는 뷰가 아니라서 popOver로 띄울 얼럿의 sourceView를 정해줄 수 없었어요)
2. 에러 throw하기 -> 에러를 던지는 함수로 바꿀 수 없어 실패
3. print로 에러 출력하기

위와 같은 이유로 fatalError를 내는 대신 3번으로 에러처리를 바꿔주었습니다.


아래는 저희가 시도해본 얼럿을 띄우는 코드 에요.
```swift
if let error = error {
    let alert = UIAlertController(title: "에러발생", message: "알수없는 에러가 발생했습니다.", preferredStyle: .actionSheet)
    let ok = UIAlertAction(title: "확인", style: .default, handler: nil)
    alert.addAction(ok)
    
    if let popOver = alert.popoverPresentationController {
        popOver.sourceView = splitViewController.view
        popOver.sourceRect = CGRect(x: splitViewController.view.bounds.midX,
                                    y: splitViewController.view.bounds.midY,
                                    width: 0,
                                    height: 0)
        popOver.permittedArrowDirections = []
    }
    present(alert, animated: true)
}
```

soo941226 replied:
넵넵 프린트로 에러 출력하는 것도 괜찮습니다👍 제 생각에 일반적으로 발생하는 에러케이스는 보통 개발 과정에서 잡힐 거라고 생각해서 프린트로 찍어보는 정도로 충분할 것 같아요. 그래도 제일 좋은 건 일반적이지 않은 에러케이스에 대해서도 조치가 되어서 유저에게 알려야할 것 같은데, 그러기 위해서는 생각하신 것처럼 alert을 띄우긴 해야할 것 같네요.

얼럿을 띄우는 건 delegation pattern이나 notification center등을 활용해서 충분히 가능하긴 한데요! 배보다 배꼽이 커질 것 같아서 일단은 말씀하신 프린트를 찍어보는 정도로 넘어가도 될 것 같아요~

soo941226 commented on CloudNotes/CloudNotes/Model/MemoDataManager.swift:
```diff
@@ -1,25 +1,50 @@
+    var isEmpty: Bool {
+        memos.count == 0
+    }
+}
```
이 부분은 왜 extension으로 빼셨는지 궁금하네요!

Allie-e replied:
새로운 메모를 만드는 것은 CRUD와 맥락이 다르다고 생각하여 분리해주었습니다!😊

soo941226 replied:
CRUD라는 게 뭔가요?

soo941226 commented on CloudNotes/CloudNotes/Controller/MemoDetailViewController.swift:
```diff
@@ -79,8 +79,27 @@ class MemoDetailViewController: UIViewController {
+            popoverController.barButtonItem = sender
+        }
+        present(alert, animated: true, completion: nil)
     }
```
외부에서 접근을 할 요소인가요?👀

na-young-kwon replied:
private 으로 변경해주었습니다!

soo941226 commented on CloudNotes/CloudNotes/Controller/MemoDetailViewController.swift:
```diff
@@ -101,6 +101,26 @@ class MemoDetailViewController: UIViewController {
+        alert.addAction(cancel)
+        alert.addAction(delete)
+        present(alert, animated: true)
+    }
```
외부에서 접근을 할 요소일까요?👀

na-young-kwon replied:
private 으로 변경해주었습니다!

soo941226 commented on CloudNotes/CloudNotes/Controller/MemoListViewController.swift:
```diff
@@ -103,6 +106,28 @@ extension MemoListViewController: UITableViewDelegate {
+        }
+        present(activityViewController, animated: true)
+    }
+    
```
지금은 상관이 없지만, 앱이 아주 크게 확장되었을 때 마찬가지의 형태를 가지고 있다면, MemoListViewController는 단독으로는 재사용하기 힘들 것 같아요. 반드시 SplitViewController와 함께 쓰여야겠네요🤔

siwonkim0 replied:
확장성을 생각한다면 그런 문제도 있을 것 같네요....! 🧐

soo941226 commented on CloudNotes/CloudNotes/Controller/MemoListViewController.swift:
```diff
@@ -141,12 +142,40 @@ extension MemoListViewController: UITableViewDelegate {
+//            try MemoDataManager.shared.viewContext.save()
+//        } catch {
+//            print("error")
+//        }
```
lastModified는 사용을 하지 않는 것으로 보이네요!

또 주석은 삭제가 된 뒤에 커밋이 되면 좋을 거 같아요.

Allie-e replied:
이후 커밋에서 lastModified도 사용하도록 수정해주었습니다!

주석은 미처 삭제하지 못하고 커밋해버렸네요ㅠㅠ 신경쓰도록 하겠습니다!

soo941226 commented on CloudNotes/CloudNotes/Model/MemoDataManager.swift:
```diff
@@ -30,6 +30,55 @@ class MemoDataManager {
+            return nil
+        }
+        return newData
+    }
```
`@discardableResult`를 사용하신 이유가 있을까요? 특히 fetch에 이게 붙어 있는 게 많이 생소하네요

siwonkim0 replied:
처음에 메서드를 생성할 때 fetch시 fetch해온 데이터를 사용하지 않는 경우도 있을수도 있겠다는 생각에 @discardableResult 키워드를 붙여주었는데, 현재 MemoListViewController에서 한번만 fetch해온 데이터를 MemoDataManager.shared.memos 에 할당하고 있기 때문에 필요없는 키워드인것 같아서 삭제하겠습니다...!

soo941226 commented on CloudNotes/CloudNotes/Controller/MemoListViewController.swift:
```diff
@@ -96,14 +96,18 @@ extension MemoListViewController: UITableViewDelegate {
+        let deletedMemo = MemoDataManager.shared.memos[indexPath.row]
         MemoDataManager.shared.memos.remove(at: indexPath.row)
         tableView.deleteRows(at: [indexPath], with: .fade)
+        MemoDataManager.shared.deleteMemo(deletedMemo)
```
`MemoDataManager.shared.memos.remove(at: indexPath.row)`
`MemoDataManager.shared.deleteMemo(deletedMemo)`
내용이 중복이 되는 게 아닌가 하는 생각이 드네요. 중복이 아니라면 기능이 분산되어있는 거 같아요. 이 부분은 개선이 되면 좋겠습니다!

na-young-kwon replied:
수박의 말씀들 듣고 저희도 기능이 분산되어 있다고 생각하여
MemoDataManager에서 배열에서 메모 삭제 + 코어데이터에서 삭제를 해주는 방식으로 수정했습니다.

따라서 MemoListViewController에서는 이 하나의 메서드로 메모를 삭제해주고 있습니다.
```swift
MemoDataManager.shared.deleteMemo(id: deletedMemo.id)
```

soo941226 commented on CloudNotes/CloudNotes/Controller/MemoListViewController.swift:
```diff
@@ -13,19 +13,20 @@ class MemoListViewController: UIViewController {
-            tableView.selectRow(at: IndexPath(row: 0, section: 0), animated: false, scrollPosition: .top)
-            tableView(tableView, didSelectRowAt: IndexPath(row: 0, section: 0))
-        }
+        setupRowSelection()
```
지금도 너무 좋은데요!!!😄👍

`MemoDataManager.shared.memos = MemoDataManager.shared.fetchMemos()`
이러한 기능이 MemoListViewController가 하는 게 말은 되는 것 같지만, MemoListViewController에서 MemoDataManager.shared.memos라고 하는 값에 어떤 기능의 결과물을 할당하는 건 코드가 복잡해지고 관리가 힘들어질 것 같아요. MemoListViewController가 직접 관리를 하게 되는 건 피하시면 좋을 거 같아요

또 setupTableView와 setupTableViewLayout가 나란히 호출되고 있는데, 이 부분만 보면 setupTableView와 setupTableViewLayout가 뭐가 다르지? 라는 생각이 먼저 드네요. 네이밍에 대해서 조금만 더 고민을 해주시면 정말 좋을 거 같습니다!

Allie-e replied:
해당 작업을 앱이 실행될 때 한번만 실행되도록 AppDelegate로 위치를 옮겨주었습니다. 추가적으로 MemoDataManager의 memos에 접근제어자도 설정해주었습니다.

setupTableView의 네이밍이 다소 모호하다는 생각이 들어 고민을 하다가 setupTableViewInitialSetting()으로 해주었는데 적절한 네이밍인지 모르겠습니다ㅎㅎ...
더 나은 네이밍이 있을까요??

soo941226 replied:
> 해당 작업을 앱이 실행될 때 한번만 실행되도록 AppDelegate로 위치를 옮겨주었습니다. 추가적으로 MemoDataManager의 memos에 접근제어자도 설정해주었습니다.

이 부분은 너무 좋은데, [아카이빙된 문서](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/CoreData/IntegratingCoreData.html#//apple_ref/doc/uid/TP40001075-CH9-SW1)에 따르면 긴 시간이 걸리는 동작을 앱실행시에 하게 되면 운영체제가 앱을 종료시킨다는 내용이 있던 걸 기억해서 다른 곳에서 하시는 게 좋지 않을까 하는 생각이 드네요

>setupTableView의 네이밍이 다소 모호하다는 생각이 들어 고민을 하다가 setupTableViewInitialSetting()으로 해주었는데 적절한 네이밍인지 모르겠습니다ㅎㅎ...
>더 나은 네이밍이 있을까요??

음 일단 setupTableView와 setupTableViewInitialSetting은 그 뜻이 크게 다르지 않은 것 같긴 하네요. 저였다면 setupTableVIewDelegatesAndCell 이라고 했을 것 같아요!

Allie-e replied:
수박이 말씀해주신 네이밍이 적절한 것 같아 반영했습니다!

https://github.com/na-young-kwon/ios-cloud-notes/commit/bcf52740a904399035e9018d36763efd3f6e574a

soo941226 commented on CloudNotes/CloudNotes/Controller/MemoDetailViewController.swift:
```diff
@@ -19,13 +19,11 @@ final class MemoDetailViewController: UIViewController {
         view.backgroundColor = .systemBackground
-        setupTextView()
-        setupNavigationItem()
-        if !MemoDataManager.shared.isEmpty {
-            memoDetailViewController(showTextViewWith: MemoDataManager.shared.memos[0])
-        }
         textView.delegate = self
         setUpNotification()
+        setupTextView()
+        setupRowSelection()
+        setupNavigationItem()
     }
```
이 부분에서 
`textView.delegate = self`
`setupTextView()`
이 부분이 공존하고 있는 게 눈에 밟히네요👀
세분은 어떻게 생각하시나요?

Allie-e replied:
listView에서 메소드 분리해줬던 것처럼 detailView에서도 메소드 분리해주었습니다!!ㅎㅎ

soo941226 commented:
안녕하세요 엘리, 제인, 조이!

말씀하신 시간보다 훨씬 빨리 PR을 보내주셔서 깜짝 놀랐네요. 정말 열심히 하시고 그만큼의 결과물이 나온 것 같아서 감탄이 많이 나왔습니다. 정말 대단하시네요👍🥳🎉

기능구현은 정말 완벽하다는 생각이 들만큼 잘 동작을 하는 것을 확인했습니다. 트러블슈팅 또한 발생한 문제의 원인과 결과를 잘 분석하셔서 해결을 해오신 것으로 보았습니다. 또 그러한 내용들을 리드미에 잘 녹여내셔서 기록을 하신 걸 확인했어요. 보면서 정말 감탄을 많이 했습니다!

그리고 한가지 고민해보시면 좋겠다는 생각이 들었는데요. Xcode를 통해서 바라보는 뷰의 계층구조가 어떻게 표현되는지 고민해보시면 좋을 거 같아요. 계층 속에서 누가 부모이고 누가 자식이고 이러한 관계를 파악해보시면서, UINavigationController, UISplitVIewController, UIViewController등이 어떤 상속 관계를 가지고 있는지도 파악해보시면 좋을 것 같아요.
제가 이러한 말을 드리는 이유는 현재의 MemoListViewController와 MemoDetailViewController가 다소 많은 일을 하는 반면에, SplitViewController는 활용이 되고 있지 않다고 느껴졌기 때문인데요.
뷰의 계층구조를 고려하시면서 이 세가지 타입이 적절히 책임을 분배하여 가지고 있다면, MemoListViewController와 MemoDetailViewController가 재사용하기 좋아질 것 같아요.

또 테스트가 필요했다면 코드를 쓰고 삭제 하는 게 아니라, 테스트 프로젝트를 활용하셨으면 코드에 대한 신뢰가 더 높아질 수 있지 않았을까 하는 아쉬움이 남습니다.

따로 질문을 주신 내용은 없어서 길게 쓸 내용은 없는 것 같네요. 각각의 커밋에 대해서 의문이 들었던 부분은 코멘트를 남겼습니다!

na-young-kwon committed:
refactor: SplitViewController 리팩토링 #7

- 상속이 불필요하기 때문에 final 클래스로 변경
- setupChildView -> assignChlidViewController로 메서드명 수정

siwonkim0 committed:
refactor: Memo 타입의 subtitle 연산 프로퍼티 기능 분리 #7

Allie-e committed:
refactor: custom splitViewController 삭제  #7

- splitVC 초기설정 sceneDelegate에서 설정
- MemoDetailViewController 내부 강제 언래핑 수정

na-young-kwon committed:
refactor: hideKeyboard -> registerGestureRecognizer로 메서드명 변경 #7

siwonkim0 committed:
refactor: 데이터가 있는지 체크하여 초기 화면 세팅하는 기능 한번에 관리 #7

- listViewController에서 MemoDataManager.shared.isEmpty 가 아닌 경우 (데이터가 있는 경우) 0번째 셀을 선택하는 것과, 0번째 데이터를 memoDetail에 보여주는 기능을 한번에 관리하도록 수정

Allie-e committed:
refactor: deleteMemo() 메소드 내부 로직 변경  #7

- 메모를 지울 때, MemoDataManager의 memos와 viewContexxt에서 각각 삭제해주던
방식을 하나의 메소드 내부에서 처리하도록 수정
- 필요없는 attribute keyword 삭제
- loadPersistentContainer에서 error 발생시 에러 메세지 프린트하도록 수정

na-young-kwon committed:
refactor: memos의 초기값을 세팅하는 메서드의 호출위치 변경 #7

- memos배열의 초기값을 세팅하는 setMemosArray 메서드 추가
- appDelegate에서 didFinishLaunchingWithOptions 시점에 호출
- memos private으로 변경
- MemoListViewController에서 테이블뷰 초기세팅을 담당하는
 setupTableViewInitialSetting 메서드 생성
- MemoDetailViewController에서 textView delegate를 설정하는 로직
 메서드로 분리

soo941226 commented:
다시 코멘트 달아주신 부분 중 상호간의 의문이 깔끔하게 해결된 부분에 대해서는 resolve를 해두었습니다!

일단 남은 기간동안 현재 코드를 개선하시는 방향으로 결정들을 하셨기 때문에 코드가 더 많이 바뀔 것이라고 생각이 되는데요. 제가 다시 코멘트를 남긴 부분은 지금 당장 고치는 게 아니라, 일단은 리팩토링을 해보시고 그 뒤에 다시 한번 고민을 해보시는 쪽으로 진행하시는 게 조금 더 공부하시기 수월할 것 같네요!

이왕 마음을 먹은 거 Delegation pattern, UISplitViewController, CoreData를 한 번 정복해본다는 생각으로 임해보시면 좋을 것 같습니다👍 저도 정복을 못했기는 했지만요😅

파이팅~!!!💪🔥

Allie-e committed:
refactor: MemoDataManager의 memos 접근제어 삭제 #7
